### PR TITLE
[shopsys] added statistics about new customers, orders and orders value on administration dashboard

### PR DESCRIPTION
**Pull request by TomasLudvik**
| Q             | A
| ------------- | ---
|Description, reason for the PR| As administrator I want to see basic stats of my store to easily check if everything is OK.
|New feature| Yes 
|[BC breaks](https://github.com/shopsys/shopsys/blob/master/docs/contributing/backward-compatibility-promise.md)| No 
|Fixes issues| ... 
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/framework/src/Model/Statistics/StatisticsRepository.php [hunk 1]
```diff
@@ -83,4 +86,76 @@ function (array $item) {
             $query->getResult()
         );
     }
+
+    /**
+     * @param \DateTimeImmutable $startDateTime
+     * @param \DateTimeImmutable $endDateTime
+     * @return int
+     */
+    public function getNewCustomersCountBetweenDates(DateTimeImmutable $startDateTime, DateTimeImmutable $endDateTime)
```
Comment (boris-brtan) [inline packages/framework/src/Model/Statistics/StatisticsRepository.php]: dont you want to use here strict-types in new methods and parts ??

File: packages/framework/src/Model/Statistics/StatisticsRepository.php [hunk 2]
```diff
@@ -83,4 +86,76 @@ function (array $item) {
...
+     * @param \DateTimeImmutable $startDateTime
+     * @param \DateTimeImmutable $endDateTime
+     * @return int
+     */
+    public function getOrdersValueBetweenDates(DateTimeImmutable $startDateTime, DateTimeImmutable $endDateTime)
+    {
+        $resultSetMapping = new ResultSetMapping();
+        $resultSetMapping->addScalarResult('total_price', 'total_price');
+
+        $query = $this->em->createNativeQuery(
```
Comment (boris-brtan) [inline packages/framework/src/Model/Statistics/StatisticsRepository.php]: is it better to use NativeQueries for this?

File: packages/framework/src/Model/Statistics/StatisticsRepository.php [hunk 3]
```diff
@@ -83,4 +86,76 @@ function (array $item) {
...
     }
+
+    /**
+     * @param \DateTimeImmutable $startDateTime
+     * @param \DateTimeImmutable $endDateTime
+     * @return int
+     */
+    public function getNewCustomersCountBetweenDates(DateTimeImmutable $startDateTime, DateTimeImmutable $endDateTime)
+    {
+        $resultSetMapping = new ResultSetMapping();
```
Comment (boris-brtan) [inline packages/framework/src/Model/Statistics/StatisticsRepository.php]: don't you thinks that many lines are identical?

File: packages/framework/src/Model/Statistics/StatisticsFacade.php [hunk 4]
```diff
@@ -74,4 +77,58 @@ public function getNewOrdersCountByDayInLastTwoWeeks()
 
         return $valueByDateTimeDataPoints;
     }
+
+    /**
+     * @param int $fromDays
+     * @param int|null $toDays
+     * @return int
+     */
+    public function getNewCustomersCount(int $fromDays, ?int $toDays = null)
```
Comment (boris-brtan) [inline packages/framework/src/Model/Statistics/StatisticsFacade.php]:
```suggestion
    public function getNewCustomersCount(int $fromDays, ?int $toDays = null): int
```
why not?

File: packages/framework/src/Controller/Admin/DefaultController.php [hunk 5]
```diff
@@ -147,4 +179,20 @@ protected function addWarningMessagesOnDashboard(): void
             );
         }
     }
+
+    /**
+     * @param int $previous
+     * @param int $current
+     * @return int
+     */
+    protected function getTrendDifference(int $previous, int $current)
```
Comment (boris-brtan) [inline packages/framework/src/Controller/Admin/DefaultController.php]: what about 
```suggestion
    protected function getTrendDifference(int $previous, int $current): int
```

File: packages/framework/src/Model/Statistics/StatisticsFacade.php [hunk 6]
```diff
@@ -74,4 +77,58 @@ public function getNewOrdersCountByDayInLastTwoWeeks()
...
+        $endDateTime = $this->getToDateTime($toDays);
+
+        return $this->statisticsRepository->getOrdersValueBetweenDates($startDateTime, $endDateTime);
+    }
+
+    /**
+     * @param int|null $toDays
+     * @return \DateTimeImmutable
+     */
+    protected function getToDateTime(?int $toDays): DateTimeImmutable
```
Comment (boris-brtan) [inline packages/framework/src/Model/Statistics/StatisticsFacade.php]: here is return type, why?

Comment (boris-brtan) [reply packages/framework/src/Model/Statistics/StatisticsFacade.php]: so i am forced to decide whether to call it with null or int?
why cant i call it without parameter?

File: packages/framework/src/Model/Statistics/StatisticsRepository.php [hunk 7]
```diff
@@ -67,14 +69,15 @@ public function getNewOrdersCountByDayBetweenTwoDateTimes(DateTime $start, DateT
         $query = $this->em->createNativeQuery(
             'SELECT DATE(o.created_at) AS date, COUNT(o.created_at) AS count
             FROM orders o
-            WHERE o.created_at BETWEEN :start_date AND :end_date
+            WHERE o.created_at BETWEEN :start_date AND :end_date AND o.status_id != :canceled
```
Comment (boris-brtan) [inline packages/framework/src/Model/Statistics/StatisticsRepository.php]: if this is some fix, can you please put it into separate commit?

File: packages/framework/src/Resources/styles/admin/components/box/overview.less [hunk 8]
```diff
@@ -60,6 +61,14 @@
                     transform: rotate(180deg);
                 }
             }
+
+            &--same {
+                background-color: @box-overview-color-same;
+
+                .svg {
+                    display: none;
```
Comment (boris-brtan) [inline packages/framework/src/Resources/styles/admin/components/box/overview.less]: what if we just not generate it via twig???

File: packages/framework/src/Model/Statistics/StatisticsRepository.php [hunk 9]
```diff
@@ -83,4 +86,76 @@ function (array $item) {
...
+     */
+    public function getOrdersCountBetweenDates(DateTimeImmutable $startDateTime, DateTimeImmutable $endDateTime)
+    {
+        $resultSetMapping = new ResultSetMapping();
+        $resultSetMapping->addScalarResult('count', 'count');
+
+        $query = $this->em->createNativeQuery(
+            'SELECT COUNT(o.created_at) AS count
+            FROM orders o
+            WHERE o.created_at BETWEEN :start_date AND :end_date AND o.status_id != :canceled',
```
Comment (boris-brtan) [inline packages/framework/src/Model/Statistics/StatisticsRepository.php]: will doctrine or postgres optimize this?

File: packages/framework/src/Controller/Admin/DefaultController.php [hunk 10]
```diff
@@ -14,6 +14,8 @@
 
 class DefaultController extends AdminBaseController
 {
+    protected const PREVIOUS_DAYS_TO_LOAD_STATISTICS_FOR = 7;
```
Comment (boris-brtan) [inline packages/framework/src/Controller/Admin/DefaultController.php]: it should be X days. Why the decision was set to just static 7 ?
is that somehow documented, or it is not needed?

File: packages/framework/src/Controller/Admin/DefaultController.php [hunk 11]
```diff
@@ -147,4 +179,20 @@ protected function addWarningMessagesOnDashboard(): void
             );
         }
     }
+
+    /**
+     * @param int $previous
+     * @param int $current
+     * @return int
```
Comment (boris-brtan) [inline packages/framework/src/Controller/Admin/DefaultController.php]: why not to round it in the twig ?

File: packages/framework/src/Controller/Admin/DefaultController.php [hunk 12]
```diff
@@ -147,4 +179,20 @@ protected function addWarningMessagesOnDashboard(): void
...
+     */
+    protected function getTrendDifference(int $previous, int $current)
+    {
+        if ($previous === 0) {
+            $trend = 100;
+        } else {
+            $trend = (int)round(($current / $previous) * 100 - 100);
```
Comment (boris-brtan) [inline packages/framework/src/Controller/Admin/DefaultController.php]:
```suggestion
            $trend = (int)round(($current / $previous - 1) * 100);
```

File: project-base/src/Shopsys/ShopBundle/DataFixtures/Demo/UserDataFixture.php [hunk 13]
```diff
@@ -66,7 +66,7 @@ public function load(ObjectManager $manager)
         $customersData = $this->loaderService->getCustomersDataByDomainId(Domain::FIRST_DOMAIN_ID);
 
         foreach ($customersData as $customerData) {
-            $customerData->userData->createdAt = $this->faker->dateTimeBetween('-1 week', 'now');
+            $customerData->userData->createdAt = $this->faker->dateTimeBetween('-2 week', 'now');
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/src/Shopsys/ShopBundle/DataFixtures/Demo/UserDataFixture.php]: You do not need to regenerate data. There is no reason for that.

File: packages/framework/src/Model/Statistics/StatisticsRepository.php [hunk 14]
```diff
@@ -83,4 +86,76 @@ function (array $item) {
...
+    public function getOrdersValueBetweenDates(DateTimeImmutable $startDateTime, DateTimeImmutable $endDateTime)
+    {
+        $resultSetMapping = new ResultSetMapping();
+        $resultSetMapping->addScalarResult('total_price', 'total_price');
+
+        $query = $this->em->createNativeQuery(
+            'SELECT SUM(o.total_price_with_vat * c.exchange_rate) AS total_price
+            FROM orders o, currencies c
+            WHERE o.created_at BETWEEN :start_date AND :end_date AND o.status_id != :canceled
+            AND o.currency_id = c.id ',
```
Comment (boris-brtan) [inline packages/framework/src/Model/Statistics/StatisticsRepository.php]:
```suggestion
            AND o.currency_id = c.id',
```

File: packages/framework/src/Resources/styles/admin/components/box/overview.less [hunk 15]
```diff
@@ -5,6 +5,7 @@
 @box-overview-number-font-size: 60px;
 @box-overview-color-up: #1ca448;
 @box-overview-color-down: #ba2929;
+@box-overview-color-same: #fcbe46;
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply packages/framework/src/Resources/styles/admin/components/box/overview.less]: I did

File: project-base/src/Shopsys/ShopBundle/DataFixtures/Demo/OrderDataFixture.php [hunk 16]
```diff
@@ -86,7 +86,7 @@ public function load(ObjectManager $manager)
         $orderData->deliveryAddressSameAsBillingAddress = true;
         $orderData->domainId = Domain::FIRST_DOMAIN_ID;
         $orderData->currency = $this->getReference(CurrencyDataFixture::CURRENCY_CZK);
-        $orderData->createdAt = $this->faker->dateTimeBetween('-1 week', 'now');
+        $orderData->createdAt = $this->faker->dateTimeBetween('-2 week', 'now');
```
[the remark this answers is not in the document]
Comment (TomasLudvik) [reply project-base/src/Shopsys/ShopBundle/DataFixtures/Demo/OrderDataFixture.php]: added

File: packages/framework/src/Model/Statistics/StatisticsRepository.php (hunk 2 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Statistics/StatisticsRepository.php]: Well yeah because you do not need any magic of entities, you just need to select raw data.

File: packages/framework/src/Model/Statistics/StatisticsRepository.php (hunk 1 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Statistics/StatisticsRepository.php]: I added strict types

File: packages/framework/src/Model/Statistics/StatisticsRepository.php (hunk 3 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Statistics/StatisticsRepository.php]: Well they are similar, but no identical... I think this is not DRY violation.

File: packages/framework/src/Model/Statistics/StatisticsFacade.php (hunk 4 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Statistics/StatisticsFacade.php]: added

File: packages/framework/src/Controller/Admin/DefaultController.php (hunk 5 above)
Comment (TomasLudvik) [reply packages/framework/src/Controller/Admin/DefaultController.php]: added

File: packages/framework/src/Model/Statistics/StatisticsFacade.php (hunk 6 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Statistics/StatisticsFacade.php]: fixed

File: packages/framework/src/Model/Statistics/StatisticsRepository.php (hunk 7 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Statistics/StatisticsRepository.php]: removed

File: packages/framework/src/Resources/styles/admin/components/box/overview.less (hunk 8 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/styles/admin/components/box/overview.less]: Well then you would need some conditions in current macro ... I prefer this way

File: packages/framework/src/Model/Statistics/StatisticsRepository.php (hunk 9 above)
Comment (TomasLudvik) [reply packages/framework/src/Model/Statistics/StatisticsRepository.php]: I do not think so

File: packages/framework/src/Controller/Admin/DefaultController.php (hunk 10 above)
Comment (TomasLudvik) [reply packages/framework/src/Controller/Admin/DefaultController.php]: I chose this because it seemed to be nice interval. This number can be changed from extending class. I would leave the decision about this to BV.

File: packages/framework/src/Controller/Admin/DefaultController.php (hunk 11 above)
Comment (TomasLudvik) [reply packages/framework/src/Controller/Admin/DefaultController.php]: Well I can typehint it this way and by design there is no expectation of such precision.

File: packages/framework/src/Controller/Admin/DefaultController.php (hunk 12 above)
Comment (TomasLudvik) [reply packages/framework/src/Controller/Admin/DefaultController.php]: done

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js [hunk 17]
```diff
@@ -51,6 +53,13 @@
                     }
                 }
 
+                // Preserve spaces if input had spaces
```
Comment (boris-brtan) [inline packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]:
```suggestion
                // Preserve spaces if input contains spaces
```

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js [hunk 18]
```diff
@@ -28,7 +28,9 @@
             var num = $this.attr('data-value');
             var nums = [num];
             var isComma = /[0-9]+,[0-9]+/.test(num);
+            var isSpace = /\s/.test(num);
```
Comment (boris-brtan) [inline packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]:
```suggestion
            var isSpace = /[0-9]+\s+[0-9]+/.test(num);
```

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js [hunk 19]
```diff
@@ -28,7 +28,9 @@
...
             num = num.replace(/,/g, '');
+            num = num.replace(/\s/g, '');
```
Comment (boris-brtan) [inline packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]: can you do it on line 32 ? or is this because of diff because the external library ???

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js [hunk 20]
```diff
@@ -51,6 +53,13 @@
                     }
                 }
 
+                // Preserve spaces if input had spaces
+                if (isSpace) {
+                    while (/(\d+)(\d{3})/.test(newNum.toString())) {
```
Comment (boris-brtan) [inline packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]: we can use [0-9] for consistency, but i like `\d` more.
i dont think this will work for higher numbers like `23 000 000`

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js [hunk 21]
```diff
@@ -51,6 +53,13 @@
...
+                        newNum = newNum.toString().replace(/(\d+)(\d{3})/, '$1'+' '+'$2');
```
Comment (boris-brtan) [inline packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]: can you use `matchAll` and `join(' ')` instead?

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js (hunk 18 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]: I am just figuring out, if the string contains space, so this is not necessary

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js (hunk 19 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]: Yes I do not want to change previous code

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js (hunk 20 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]: This works, tested on milions

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js (hunk 21 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]: I am consistent with previous code

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js (hunk 17 above)
Comment (TomasLudvik) [reply packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]: fixed

File: packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js [hunk 22]
```diff
@@ -51,6 +53,13 @@
                     }
                 }
 
+                // Preserve spaces if input contained spaces
+                if (isSpace) {
+                    while (/(\d+)(\d{3})/.test(newNum.toString())) {
+                        newNum = newNum.toString().replace(/(\d+)(\d{3})/, '$1'+' '+'$2');
```
Comment (boris-brtan) [inline packages/framework/src/Resources/scripts/admin/plugins/jquery.counterup.js]:
```suggestion
                        newNum = newNum.toString().replace(/(\d+)(\d{3})/, '$1 $2');
```

**Comment by TomasLudvik** [comment]
> Nice job. I love it!
> Two things:
> 
> * I would like to know how you updated the demodata
> * I need to display the time period for those statistics. Is it a week? I will discuss with @tomshopsys how to display it properly.

Hello @LukasHeinz,

* I just changed random date for last week to random date for two last weeks.
* Well I think this is part of redesigning so we can merge this to @tomshopsys branch and he can update it later.

File: packages/framework/src/Model/Statistics/StatisticsFacade.php [hunk 23]
```diff
@@ -85,9 +85,11 @@ public function getNewOrdersCountByDayInLastTwoWeeks(): array
      */
     public function getNewCustomersCount(int $fromDays, ?int $toDays = null): int
     {
-        $startDateTime = new DateTimeImmutable('- ' . $fromDays . ' days midnight');
+        $startDateTime = new DateTimeImmutable('- ' . $fromDays . ' days');
         $endDateTime = $this->getToDateTime($toDays);
 
+        $startDateTime->getTimestamp()
```
Comment (boris-brtan) [inline packages/framework/src/Model/Statistics/StatisticsFacade.php]:
```suggestion
        $startDateTime->getTimestamp()
```
